services/wireguard/connection/connection.go

```diff
@@ -223,29 +228,32 @@ func (c *Connection) isNoopPinger() bool {
 
 // Stop stops wireguard connection and closes connection endpoint.
 func (c *Connection) Stop() {
-	log.Info().Msg("Stopping WireGuard connection")
-	c.stateCh <- connection.Disconnecting
-	c.sendStats()
-
-	if c.connectionEndpoint != nil {
-		if err := cleanDNS(c.configDir, c.connectionEndpoint.InterfaceName()); err != nil {
-			log.Error().Err(err).Msg("Failed to clear DNS")
+	c.stopOnce.Do(func() {
+		log.Info().Msg("Stopping WireGuard connection")
+		c.stateCh <- connection.Disconnecting
+
+		if c.connectionEndpoint != nil {
+			// c.sendStats()
+
+			if err := c.dnsManager.Clean(c.opts.DNSConfigDir, c.connectionEndpoint.InterfaceName()); err != nil {
+				log.Error().Err(err).Msg("Failed to clear DNS")
+			}
+			if err := c.connectionEndpoint.Stop(); err != nil {
+				log.Error().Err(err).Msg("Failed to close wireguard connection")
+			}
 		}
-		if err := c.connectionEndpoint.Stop(); err != nil {
-			log.Error().Err(err).Msg("Failed to close wireguard connection")
+
+		if c.removeAllowedIPRule != nil {
```

Comment:
in constructor, you could set this to func() {}

Then you would not need this check. You would also be able to do the following:
```golang
defer  c.removeAllowedIPRule()
if c.connectionEndpoint == nil {
return
}

if err := c.dnsManager.Clean(c.opts.DNSConfigDir, c.connectionEndpoint.InterfaceName()); err != nil {
				log.Error().Err(err).Msg("Failed to clear DNS")
			}
			if err := c.connectionEndpoint.Stop(); err != nil {
				log.Error().Err(err).Msg("Failed to close wireguard connection")
			}

Reply:
This looks more clear as it shows that `removeAllowedIPRule` was actually applied on Start().
Also I can't return if connectionEndpoint is nil as I need to close all channels at the end.